Add Authenticator config `allow_all` and `allow_existing_users` (#4701)
jupyterhub/jupyterhub: docs/source/reference/authenticators.md

appear in jupyterhub help output and auto-generated configuration files
via `jupyterhub --generate-config`.

(authenticator-allow)=

### Allowing access

When dealing with logging in, there are generally two _separate_ steps:

authentication
: identifying who is trying to log in, and

authorization
: deciding whether an authenticated user is allowed to access your JupyterHub

{meth}`Authenticator.authenticate` is responsible for authenticating users.
It is perfectly fine in the simplest cases for `Authenticator.authenticate` to be responsible for authentication _and_ authorization,
in which case `authenticate` may return `None` if the user is not authorized.

> However, Authenticators also have have two methods {meth}`~.Authenticator.check_allowed` and {meth}`~.Authenticator.check_blocked_users`, which are called after successful authentication to further check if the user is allowed.

Reviewer: ```suggestion
However, Authenticators also have two methods, {meth}`~.Authenticator.check_allowed` and {meth}`~.Authenticator.check_blocked_users`, which are called after successful authentication to further check if the user is allowed.
```

If `check_blocked_users()` returns False, authorization stops and the user is not allowed.

If `Authenticator.allow_all` is True OR `check_allowed()` returns True, authorization proceeds.

:::{versionadded} 5.0
{attr}`.Authenticator.allow_all` and {attr}`.Authenticator.allow_existing_users` are new in JupyterHub 5.0.

By default, `allow_all` is False,
which is a change from pre-5.0, where `allow_all` was implicitly True if `allowed_users` was empty.
:::

### Overriding `check_allowed`

:::{versionchanged} 5.0

> `check_allowed()` is **not called** is `allow_all` is True.

Reviewer: ```suggestion
`check_allowed()` is **not called** if `allow_all` is True.
```

:::

:::{versionchanged} 5.0
Starting with 5.0, `check_allowed()` should **NOT** return True if no allow config

> is specified (`allow_all` should be used instead).

Reviewer: Is there anything wrong with relying on `check_allowed()` to do everything in a custom authenticator?

Reviewer 2: Not _really_. The only real downside is that it becomes impossible to specify "nobody should be allowed." if the total absence of config results in `allow_all`. I'd stand by this statement, though, as a recommendation to Authenticators.

:::

The base implementation of {meth}`~.Authenticator.check_allowed` checks:

- if username is in the `allowed_users` set, return True
- else return False

:::{versionchanged} 5.0

> Prior to 5.0, the check was
> 
> ```python
> if (not allowed_users) or username in allowed_users:
> ```

Reviewer: ```suggestion
Prior to 5.0, this would also return True if `allowed_users` was empty.
```

> but the implicit `not allowed_users` has been replaced by explicit `allow_all`, which is checked _before_ calling `check_allowed`.

Reviewer: ```suggestion
For clarity this is no longer the case. A new `allow_all` property (default False) has been added which is checked _before_ calling `check_allowed`.
```

> `check_allowed` **is not called** if `allow_all` is True.

Reviewer: ```suggestion
If `allow_all` is True this takes priority over `check_allowed` which will be ignored.
```

If your Authenticator subclass similarly returns True when no allow config is defined,
this is fully backward compatible for your users, but means `allow_all = False` has no real effect.

You can make your Authenticator forward-compatible with JupyterHub 5 by defining `allow_all` as a boolean config trait on your class:

```python
class MyAuthenticator(Authenticator):

    # backport allow_all from JupyterHub 5

>     allow_all = Bool(False, config=True)

Reviewer: I can't remember how Traitlets works- if someone sets `c.Authenticator.allow_all = True|False` will this also set `MyAuthenticator.allow_all` if the property is redefined in the subclass?

Reviewer 2: Yes, redefining the trait will still inherit the parent config. Config is loaded if it's tagged as `config=True` on the loading class and the config section name matches any parent Configurable class. Only if MyAuthenticator redefined it without `config=True` would it not load `Authenticator.allow_all`.

            # from JupyterHub 5
            # this branch will never be taken with jupyterhub >=5
            return True
        if self.allowed_something and user_has_something(username):
            return True
        # repeat for each allow
        if allow_config and allow_met:
            return True
        # should always have this at the end
        if self.allowed_users and username in self.allowed_users:
            return True
        # do not call super!
        # super().check_allowed is not safe with JupyterHub < 5.0,
        # as it will return True if allowed_users is empty
        return False
```

Key points:

- `allow_all` is backported from JupyterHub 5, for consistent behavior in all versions of JupyterHub (optional)
- restrictive configuration is checked in `check_blocked_users`
- if any restriction is not met, `check_blocked_users` returns False
- permissive configuration is checked in `check_allowed`
- if any `allow` condition is met, `check_allowed` returns True

So the logical expression for a user being authorized should look like:

> if ALL restrictions are met AND ANY admissions are met: user is authorized

#### Custom error messages

> Any of these authentication and authorization methods may

Reviewer: ```suggestion
Any of these authentication and authorization methods may raise a `web.HTTPError` Exception
```
